xls/modules/rle/rle.x

```diff
@@ -0,0 +1,306 @@
+// to the data_r        ──────────────────╨─────╨─────╨─────╨─────╨─────╨────
+
+import std
+
+// structure containing output data from an RLE encoder
+struct EncOutData<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+    symbol: bits[SYMBOL_WIDTH],
+    count: bits[COUNT_WIDTH],
+}
+
+// structure to preserve the state of an RLE encoder for the next iteration
+pub struct RLEEncState<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+    // previous symbol and symbol count
+    prev: EncOutData<SYMBOL_WIDTH, COUNT_WIDTH>,
+    // if flush of data was requested in the previous iteration of the
+    // next function. Used to skip reading symbols from the input channel,
+    // to get time span for sending back data from the encoder.
+    flush: bool,
+
+}
+
+// RLE encoder implementation
+pub proc RLEEnc<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+    // input with symbols (should be read together with flush_r)
+    symbol_r: chan<bits[SYMBOL_WIDTH]> in;
+    // input with flush request (should be read together with symbol_r)
+    flush_r: chan<bool> in;
+    // output with data, contains symbol and symbol count
+    data_s: chan<EncOutData<SYMBOL_WIDTH, COUNT_WIDTH>> out;
+
+    init {(
+        RLEEncState<SYMBOL_WIDTH, COUNT_WIDTH> {
+            prev: EncOutData<SYMBOL_WIDTH, COUNT_WIDTH> {
+                symbol: bits[SYMBOL_WIDTH]:0,
+                count: bits[COUNT_WIDTH]:0,
+            },
+            flush: false,
+        }
+    )}
+
+    config (
+        symbol_r: chan<bits[SYMBOL_WIDTH]> in,
+        flush_r: chan<bool> in,
+        data_s: chan<EncOutData<SYMBOL_WIDTH, COUNT_WIDTH>> out,
+    ) {(symbol_r, flush_r, data_s)}
+
+    next (tok: token, state: RLEEncState<SYMBOL_WIDTH, COUNT_WIDTH>) {
+        let (flush_tok, flush) = recv_if(tok, flush_r, !state.flush, false);
+        let (symbol_tok, symbol) = recv_if(tok, symbol_r, !state.flush, bits[SYMBOL_WIDTH]:0);
+        let recv_tok = join(flush_tok, symbol_tok);
+
+        let prev_symbol_valid = state.prev.count != bits[COUNT_WIDTH]:0;
+        let symbol_differ = prev_symbol_valid && (symbol != state.prev.symbol);
+        let overflow = state.prev.count == std::unsigned_max_value<COUNT_WIDTH>();
+
+        let next_state = if (state.flush) {
+            RLEEncState {
+                prev: EncOutData<SYMBOL_WIDTH, COUNT_WIDTH> {
+                    symbol: bits[SYMBOL_WIDTH]:0,
+                    count: bits[COUNT_WIDTH]:0,
+                },
+                flush: flush,
+            }
+        } else if (symbol_differ || overflow) {
+            RLEEncState {
+                prev: EncOutData<SYMBOL_WIDTH, COUNT_WIDTH> {
+                    symbol: symbol,
+                    count: bits[COUNT_WIDTH]:1,
+                },
+                flush: flush,
+            }
+        } else {
+            RLEEncState {
+                prev: EncOutData<SYMBOL_WIDTH, COUNT_WIDTH> {
+                    symbol: symbol,
```

Comment:
even if that's equivalent, I think `state.prev.symbol` would be more readable.

Reply:
Yes, they are almost equivalent. Except from the first iteration when the `state.prev.symbol` is `0`. So actually using `symbol` does not require any logic to distinguish between the first and any other iteration.